[Enhancement]Fix Conversion warning for gcc(#12071)

srlch opened the pull request:
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #12071

## Problem Summary(Required) ：
Problem:
If we enable Wconversion for compiling in gcc, we will get a lot warning for it. This issue is mainly remove part of warning from the code.

Solution:
We will use more reasonable variable type in the code, except for:
1. The modification will cause performance drop.
2. Some small-length variables come from third-party libraries, so we can only ensure that our code is safe when we used it.
3. Some functions logically need to do this type conversion.
4. From the semantics of the function, there will be no overflow.

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

trueeyu commented on gensrc/proto/olap_file.proto:
```diff
@@ -114,8 +114,9 @@ message RowsetMetaPB {
     optional PUniqueId tablet_uid = 21;
-    // total number of segments
-    optional int64 num_segments = 22;
+    // total number of segments, it is nearly impossible that the num_segments will
```
There will be compatibility issues after the change

trueeyu commented on be/src/storage/short_key_index.cpp:
```diff
@@ -31,7 +31,7 @@ using strings::Substitute;
 namespace starrocks {
 
 Status ShortKeyIndexBuilder::add_item(const Slice& key) {
-    put_varint32(&_offset_buf, _key_buf.size());
```
There may be compatibility issues

srlch committed:
[BugFix]Fix Conversion warning for gcc(#12071)

Problem:
If we enable Wconversion for compiling in gcc, we will get a lot warning for it. This issue is mainly remove part of warning from the code.

Solution:
We will use more reasonable variable type in the code, except for:
1. The modification will cause performance drop.
2. Some small-length variables come from third-party libraries, so we can only ensure that our code is safe when we used it.
3. Some functions logically need to do this type conversion.
4. From the semantics of the function, there will be no overflow.

jaogoy commented on be/src/storage/tablet_meta.h:
```diff
@@ -116,7 +116,7 @@ class TabletMeta {
     void set_tablet_id(int64_t tablet_id) { _tablet_id = tablet_id; }
     int32_t schema_hash() const;
-    int16_t shard_id() const;
+    int32_t shard_id() const;
```
We'd better define `typedef int32_t ShardID`, and use ShardID only.

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
run starrocks_admit_test
